src/protocol/boolean/carries.rs

```diff
@@ -0,0 +1,400 @@
+        let s = Self::step1(a, b, ctx.narrow(&Step::AMultB), record_id).await?;
+        let e = Self::step2(a, b, &s, ctx.clone(), record_id).await?;
+        let f = Self::step3(&e, ctx.narrow(&Step::CarryPropagation), record_id).await?;
+        let s = f.iter().map(|&x| x.s).collect::<Vec<_>>();
+        Ok(s)
+    }
+
+    /// Step 1. for `i=0..l=1`, `[s_i] = MULT([a_i], [b_i])`
+    ///
+    /// The interpretation is, `s_i` = 1 iff i'th bit of a and b are 1.
+    ///
+    /// # Example
+    /// ```ignore
+    ///   // Input
+    ///   [a] = 1 1 0 1 0 1 1 0   // 214 in LE
+    ///   [b] = 0 0 0 1 1 1 0 0   // 28 in LE
+    ///   // Output
+    ///   [s] = 0 0 0 1 0 1 0 0
+    /// ```
+    async fn step1<F: Field>(
+        a: &[Replicated<F>],
+        b: &[Replicated<F>],
+        ctx: ProtocolContext<'_, Replicated<F>, F>,
+        record_id: RecordId,
+    ) -> Result<Vec<Replicated<F>>, BoxError> {
+        let mul = zip(repeat(ctx), zip(a, b))
+            .enumerate()
+            .map(|(i, (ctx, (&a_bit, &b_bit)))| {
+                let c = ctx.narrow(&BitOpStep::Step(i));
+                async move { c.multiply(record_id, a_bit, b_bit).await }
+            });
+        try_join_all(mul).await
+    }
+
+    /// Step 2. for `i=0..l=1`, generate a vector of tuples `([s], [p], [k])` where
+    /// `[p_i] = [a_i] + [b_i] - 2*[s_i]`, and `[k_i] = 1 - [s_i] - [p_i]`
+    ///
+    /// The interpretations is:
+    ///   * `s_i = 1` iff carry is set at position `i` (i.e. `a_i + b_i = 2`)
+    ///   * `p_i = 1` iff carry would be propagated at position `i` (i.e. `a_i + b_i = 1`)
+    ///   * `k_i = 1` iff a carry would be killed at position `i` (i.e. `a_i + b_i = 0`)
+    ///
+    /// Therefore, we compute `p_i = XOR(a_i, b_i)`, and `k_i = !OR(a_i, b_i)`
+    ///
+    /// # Example
+    /// ```ignore
+    /// // Input
+    ///   [a] = 1 1 0 1 0 1 1 0   // 214 in LE
+    ///   [b] = 0 0 0 1 1 1 0 0   // 28 in LE
+    ///   [s] = 0 0 0 1 0 1 0 0
+    /// // Output
+    ///   [s] = 0 0 0 1 0 1 0 0   // `a_i & b_i`
+    ///   [p] = 1 1 0 0 1 0 1 0   // `a_i ^ b_i`
+    ///   [k] = 0 0 1 0 0 0 0 1   // `!(a_i | b_i)`
+    /// ```
+    #[allow(clippy::many_single_char_names)]
+    async fn step2<F: Field>(
+        a: &[Replicated<F>],
+        b: &[Replicated<F>],
+        s: &[Replicated<F>],
+        ctx: ProtocolContext<'_, Replicated<F>, F>,
+        record_id: RecordId,
+    ) -> Result<Vec<CarryPropagationShares<F>>, BoxError> {
+        let xor_ctx = ctx.narrow(&Step::AXorB);
+        let or_ctx = ctx.narrow(&Step::AOrB);
+        let (p_futures, k_futures): (Vec<_>, Vec<_>) = zip(a, b)
+            .enumerate()
+            .map(|(i, (&a_bit, &b_bit))| {
+                let c1 = xor_ctx.narrow(&BitOpStep::Step(i));
+                let c2 = or_ctx.narrow(&BitOpStep::Step(i));
+                (
+                    async move { xor(c1, record_id, a_bit, b_bit).await },
```

Comment:
I get what you're doing here (this is indeed computing an XOR). But I think that this way you've unfortunately replicated the exact same multiplication from step_1. Given we are already passing in `s` this doesn't require any more communication and can be done locally.